### DISCUSSION TITLE
MIDI Library without Arduino.h

### DISCUSSION THREAD

**Discussion by zerbian** [FortySevenEffects/arduino_midi_library#209] [Q&A] [answered]
Hello there,
I am currently trying to write code without using the Arduino.h implmented feature, but I still want to use the MIDI Library in my program. Maybe someone can point me to the right resources for setting the UART / Serial register. I am currently writing for a ATMega328p, any help is welcome!
Thanks

**Comment by franky47** [comment, marked as answer]
The MIDI library can be used on AVR processors without the Arduino core.

I have some examples using a hardware abstraction layer based on C++ templates ([`FortySevenEffects/cake`](https://github.com/FortySevenEffects/cake)) at [`FortySevenEffects/monosynth`](https://github.com/FortySevenEffects/monosynth). It might need a little digging to see how it works. 

The general idea is to use a custom transport for the MIDI Library, that maps to your RX/TX buffers, themselves being connected to the UART via interrupt handlers.

**Reply by zerbian** [reply]
That was fast! Thank you!
I will try to familiarize myself with the code and inner workings, but I am more accustomed to C (which is very hands on with the chip datasheet) and not C++

**Reply by franky47** [reply]
You can also look at the default [Serial transport](https://github.com/FortySevenEffects/arduino_midi_library/blob/d9149d19df867abbf78cfcd37ef9d0e14dedca7f/src/serialMIDI.h#L42-L92), it shows the API contract expected by the library. If you duplicate the class and implement its methods by connecting them to the UART, you'll be able to use that as a transport for the library:

```c++
MyCustomTransport transport;
midi::MidiInterface<MyCustomTransport> MIDI(transport);
```

Note: because your custom transport is likely not to use [`SerialSettings`](https://github.com/FortySevenEffects/arduino_midi_library/blob/d9149d19df867abbf78cfcd37ef9d0e14dedca7f/src/serialMIDI.h#L33-L40), remember to start your UART at the MIDI baud rate of 31250.
